### PR TITLE
RD-220 Added cypress commons

### PR DESCRIPTION
**Pull request by kubama**

### CODE REVIEW

**Comment by Gelio** [comment]
This PR clashes with the changes in https://github.com/cloudify-cosmo/cloudify-ui-common/pull/46. Actually, I reckon all of the PRs related to this ticket will clash with changes for TypeScript.

In the future, I reckon it would be easier not to make such changes in parallel, or at least consult it, as adding Cypress commands to common requires special handling for TypeScript to work :worried:

**Comment by kubama** [comment]
@Gelio This PR adds new source files only, does not modify any existing source files, so I admit I didn't think it could clash.
If it does, I can wait with this until TS PRs are merged.

**Comment by Gelio** [comment]
@kubama Sure, no problem, I guess that could have been unexpected.

I'm not sure what changes you are planning for the `stage` repo, but so far, I think what will be required is exposing the typings for the custom Cypress commands that this PR introduces (`getAdminToken`) and importing them in stage, which I have not thought about yet.

Queueing this PR after TypeScript's PRs sounds good to me. This way we can focus on how to expose Cypress typings after TS works fine :slightly_smiling_face: Are you sure that's fine for you? It's a pity we both worked on configuration changes at the same time :worried:

**Comment by qooban** [comment]
@kubama After discussing with @Gelio I think that we can wait a bit with merging your PR and first introduce TypeScript support (merge https://github.com/cloudify-cosmo/cloudify-ui-common/pull/46 + release new ui-common + merge all related PRs). 

@Gelio I'll try to take a look tomorrow at #46 . if I won't be able to do detailed review till TypeScript introduction meeting tomorrow, I won't hold it on and we can finalize introducing TypeScript support.

Anyways, we'll find a way to push all changes to the master. Let's discuss a plan on tomorrow's internal meeting.

**Comment by kubama** [comment]
OK, let it be

File: cypress/cypress.json
Comment (Gelio) [inline cypress/cypress.json:7]: I'm thinking whether it shouldn't be the other way around:
1. Each repository has its own `cypress.json` that specifies things like paths, etc.
2. In the `plugins` section of each final repository (stage, composer), there is an import `cloudify-ui-common/cypress/plugins/index.js`. That `index.js` would dynamically change Cypress's configuration and apply common settings.

This is because I have not seen any way to `extend` a `cypress.json`, as we do with ESLint or TypeScript.

If we use common plugins to extend the configuration, each repository can easily override the default configuration, if need be (e.g. do not capture videos or use a different number of retries).

https://docs.cypress.io/api/plugins/configuration-api.html#Usage

So, I would imagine that we have a single plugin in this repository that only applies the shared settings. It could also automatically add the `getAdminToken` command, or we could explicitly include it by creating a new file in `stage` and `composer` and importing `cloudify-ui-common/cypress/support/getAdminToken`. 

Then, other changes in this PR would include changes to eslint.

What do you think?

Comment (kubama) [reply cypress/cypress.json:7]: You can still set/override any configuration options on per-repository basis with `-c` parameter passed to the runner. My plan was to use it, so for example in composer the `e2e` npm script would be extended to `cypress run -C node_modules/cloudify-ui-common/cypress/cypress.json -c baseUrl=http://localhost:9000`. In this case we don't need a per-repository plugins file.

Importing `cloudify-ui-common/cypress/support`, as per your proposal, is also part of the plan.

Comment (Gelio) [reply cypress/cypress.json:7]: I see :slightly_smiling_face: 

Adding `-c baseUrl=http://localhost:9000` to the command itself makes it harder to maintain, IMO, than a config file. If it was specified for a config file, then no matter how you run the command (whether it's `npm run cypress` or `npx cypress`), it would be picked up. If I run `npx cypress`, then I guess it will fail, since `cypress.json` will not be present in the repo

Comment (kubama) [reply cypress/cypress.json:7]: > having a plugin that imports some other plugin is also less magical than using a config from another repo directly

OK, I can agree on this. Will make the change.

However, when it comes to cypress.json the goal was to reuse this config file as it is identical for stage/composer, except the `baseUrl` property. This property could be set programmatically making the `-c` command line parameter unnecessary, that's good, however we still need `-C` to specify non-standard cypress.json location. Without it the runner will use default configuration and will fail as it won't have correct paths set  - so unfortunately we can't set all configuration options at runtime as part of plugin initialization.

Comment (Gelio) [reply cypress/cypress.json:7]: Hmm... I can't figure out setting which shared configuration value is unachievable with plugins and has to be done with a config. Do you have any specifics in mind? :thinking:

Comment (kubama) [reply cypress/cypress.json:7]: For sure the `pluginsFile` needs to first point to correct plugin initialization file, without it the runner will look for it in a default location, where it won't be found, so runtime configuration won't run. If so I suspect other paths may need to be set statically as well, I haven't verify it though, I'll try to check it later today.

Comment (Gelio) [reply cypress/cypress.json:7]: My intention was to have a plugins file in each repository (stage and composer), which would say:

```ts
import setCommonConfig from 'cloudify-ui-common/cypress';

module.exports = () => {
  setCommonConfig();
};
```

or something similar

Comment (kubama) [reply cypress/cypress.json:7]: Yes, I understand, but you still need to set path to this file in the static configuration, otherwise it won't be picked up as we are using non-standard directory structure.

**Review by qooban — DISMISSED** [review]
LGTM 👍🏻

**Review by Gelio — CHANGES_REQUESTED** [review]
I'm putting this review to hold this PR for now, as I thought it's still in progress, but I see @qooban approved it :sweat_smile: 

This is since we need to configure emitting TS declarations for custom commands defined here :smile: 

Also, I would still suggest another way of structuring the way configuration is shared (i.e. what I described in https://github.com/cloudify-cosmo/cloudify-ui-common/pull/47#discussion_r570786227)

**Comment by kubama** [comment]
Cypress commons are now moved to TS.
After playing with it I was not able to directly import support/plugin ts files exported by this package, so I added tsc compilation - consumers of cypress commons will now import compiled JS files - I tested it and it seems to be working just fine.

> Also, I would still suggest another way of structuring the way configuration is shared (i.e. what I described in [#47 (comment)](https://github.com/cloudify-cosmo/cloudify-ui-common/pull/47#discussion_r570786227))

This is how it is going to work. However, the `setCommonConfig` will only contain logic that's currently in stage/composer plugin.js - as I explained static cypress.json is still required, so I would like to keep all static configuration in one place instead of having some configuration options set in `cypress.json`, and some in `setCommonConfig`.

**Review by qooban — DISMISSED** [review]
LGTM 👍🏻

**Comment by qooban** [comment]
@Gelio Please review and let's try to finish RD-220 this week (merge this PR, release new cloudify-ui-common version, merge incoming PR in stage). As we already discussed with @kubama Composer part will be done in a separate task (after introducing TypeScript to Composer).

File: configs/eslint-common-cypress.json
Comment (Gelio) [inline configs/eslint-common-cypress.json:2]: I'm wondering... Should we not be extending `eslint-ts-overrides.json` here too?

That config extends `@typescript-eslint/recommended`, which I am sure will provide some useful rules

File: cypress/plugins/index.ts
Comment (Gelio) [inline cypress/plugins/index.ts:11]: This is one way of solving the problem, and probably the easier one :slightly_smiling_face: Writing custom type definitions for an external package is easy enough in this case, so I would advise:

1. Creating a `d.ts` file somewhere in this TS project (e.g. `cypress/plugins/typings.d.ts`) with the following contents:

    ```ts
    // NOTE: preexisting types were not found
    declare module '@cypress/code-coverage/task' {
        const coverageTask: Cypress.PluginConfig;
        export default coverageTask;
    }
    ```

    Notice that I reused the `Cypress.PluginConfig` type from Cypress instead of writing the function itself, which I always encourage :slightly_smiling_face: 

2. Removing the comments here :smile: `coverageTask` will have the appropriate type
3. After making this change you will see errors below, when calling it. See comment related to that code for a solution

File: cypress/plugins/index.ts
Comment (Gelio) [inline cypress/plugins/index.ts:28]: After applying the suggestion for typing `coverageTask` you will see an error when calling it, because `on` has a different type. You specified it as `Cypress.Actions`, and `Cypress.PluginConfig` expects `Cypress.PluginEvents`.

After taking a look at the type declarations for each of them and consulting [the list of plugin events in the docs](https://docs.cypress.io/api/plugins/writing-a-plugin.html#List-of-events), `Cypress.PluginEvents` is the correct type for `on` :slightly_smiling_face: 

However, we could do better :smile: Reusing types for the win here:

```suggestion
const setupPluginsAndConfig: Cypress.PluginConfig = (on, config) => {
```

A lot less code, and more type safety :slightly_smiling_face:

File: cypress/README.md
Comment (Gelio) [inline cypress/README.md:7]: Trailing space at the end of the line
```suggestion
* `cypress.json` - base Cypress runner static configuration
```

File: cypress/tsconfig.json
Comment (Gelio) [inline cypress/tsconfig.json:3]: I was thinking about using `declarationMap` and `outDir` here, to avoid keeping TS compilation output in the same directories as the source file, but that would likely require an additional `package.json`, and IMO that's too much overhead ATM.

I'll leave this comment here to let you know that I considered this option :smile: and notify you, just in case you did not think of that :slightly_smiling_face: You can resolve it if you agree that it's too much overhead

File: src/tsconfig.json
Comment (Gelio) [inline src/tsconfig.json:5]: Can we adjust this flag after we agree on that as a team? If we decide to drop `checkJs`, that would be best done in `tsconfig.base.json`, and not individual project tsconfigs :slightly_smiling_face:

File: cypress/support/index.ts
Comment (Gelio) [inline cypress/support/index.ts:26]: Is this command visible in stage's `cy`? I have not checked, I'm just asking to be sure that extending the namespace in a separate package worked :smile:

File: cypress/cypress.json
Comment (Gelio) [inline cypress/cypress.json:1]: Just to be sure and to continue the conversation from the other thread:

> However, the setCommonConfig will only contain logic that's currently in stage/composer plugin.js - as I explained static cypress.json is still required, so I would like to keep all static configuration in one place instead of having some configuration options set in cypress.json, and some in setCommonConfig.

After considering it for a while, I reckon that it makes sense to do it this way around :smile: I was worried about not being able to override the individual settings in Stage, but we could always write a custom plugin that would do it, and that plugin would take precedence over the config file :slightly_smiling_face: 

Thus, all is good here :slightly_smiling_face: I'm just posting the comment to give my thoughts

**Review by Gelio — COMMENTED** [review]
Nice work tweaking TS here :smile: :+1: looks solid

File: cypress/support/index.ts
Comment (Gelio) [reply cypress/support/index.ts:26]: And for extra effort: does jumping to the function definition work too, so we see the source code in `node_modules`?

Comment (Gelio) [reply cypress/support/index.ts:26]: Nevermind, I verified it :smile: It's working flawlessly when we 

```ts
import 'cloudify-ui-common/cypress/support';
```

so, good job! :smile:

File: configs/eslint-common-cypress.json
Comment (kubama) [reply configs/eslint-common-cypress.json:5]: I just use WebStorm's auto formatting. But if prettier says it can be prettier then let it be 😄.

File: configs/eslint-common-cypress.json
Comment (kubama) [reply configs/eslint-common-cypress.json:12]: Looks like I instinctively removed it as, again, WebStorm marked it as invalid syntax.
I brought the comment back and altered my WS configuration so it no longer reports JSON comments as errors.

File: cypress/tsconfig.json
Comment (kubama) [reply cypress/tsconfig.json:3]: Actually I thought about the same, but I think as this package grows at some point in the future its structure will need to be refactored anyway, possibly by splitting it into smaller packages. But since it's still relatively small I'm not sure how exactly it should look, so for now I'd leave it as it is.

File: src/tsconfig.json
Comment (kubama) [reply src/tsconfig.json:5]: Sorry about that - this was committed accidentally. I turned this flag off as all these (false) errors were really annoying and forgot about it. Lets discuss tomorrow. I'm reverting the change now.

File: cypress/tsconfig.json
Comment (qooban) [reply cypress/tsconfig.json:3]: Good points, guys. I'm taking that as an early call for splitting single `cloudify-ui-common` package into few packages. Makes sense. Will add it to UI Technical Debt epic.

File: configs/eslint-common-cypress.json
Comment (Gelio) [reply configs/eslint-common-cypress.json:12]: Awesome :smile:

File: src/tsconfig.json
Comment (Gelio) [reply src/tsconfig.json]: Sure, no problemo :slightly_smiling_face:

File: configs/eslint-common-cypress.json
Comment (kubama) [reply configs/eslint-common-cypress.json]: I don't think it should. This file contains only Cypress specific configuration. Once this file is extended by `.eslintrc.json` in `cypress` directory the rest of the configuration is inherited from the package's top level `.eslintrc.json`, which in case of ui-common/stage/composer extends `eslint-common.json`, which in turn extends `eslint-ts-overrides.json`. So configuration inheritance already takes care of this.

Comment (Gelio) [reply configs/eslint-common-cypress.json]: Today I learned that ESLint merges subdirectory configs with the configs from parent directories :smile: Thanks!

Resources for whoever will be looking at this thread in the future and did not know it: https://eslint.org/docs/user-guide/configuring/configuration-files#cascading-and-hierarchy

File: cypress/plugins/index.ts
Comment (kubama) [reply cypress/plugins/index.ts]: Thanks for a ready solution - done

File: cypress/plugins/index.ts
Comment (kubama) [reply cypress/plugins/index.ts]: I admit I didn't even realize it's possible to reuse a complete function definition in such a way - cool, thanks.

Comment (Gelio) [reply cypress/plugins/index.ts]: No problem :smile: That's one of the nicest features of TS that I use quite often :slightly_smiling_face:

File: cypress/tsconfig.json
Comment (Gelio) [inline cypress/tsconfig.json:3]: One minor thing:

Since this project has only TS files, we could add:

```ts
"allowJs": false
```

as a compiler option, to disallow adding JS files here and force using TS :slightly_smiling_face:

File: cypress/tsconfig.json
Comment (Gelio) [inline cypress/tsconfig.json:1]: Formatting this file with prettier yields different results, just FYI :smile: I will most likely format it with prettier (it's set as my default formatter in VSCode), so I'll be seeing some annoyance in the future

Maybe we could all use prettier as a default formatter?

File: package.json
Comment (Gelio) [inline package.json:16]: I just thought... Can we do it shorter?

```suggestion
        "cypress",
```

I think that would work equally well, as it does with `backend/typings` or `dist` or any other directory below

**Review by Gelio — DISMISSED** [review]
I see there are some lint warnings now:
image

Do you think you could address them? 

For `no-explicit-any`, you could try to replace `...args: any[]` with `...args: unknown[]`, just FYI

For `explicit-module-boundary-types`, I'm not sure :smile: You could either write that function returns `void`, or disable the rule. I'm actually thinking about disabling this rule entirely

The PR looks good, so I'm approving it :slightly_smiling_face: Only a few minor points left

File: package.json
Comment (kubama) [reply package.json]: Of course, it was just a leftover from a previous version of this line

File: cypress/tsconfig.json
Comment (kubama) [reply cypress/tsconfig.json:3]: Done

**Comment by kubama** [comment]
> Do you think you could address them?

Addressed

File: cypress/tsconfig.json
Comment (qooban) [reply cypress/tsconfig.json:1]: We already use prettier (through ESLint), maybe not in the same way, maybe some configuration should be updated, but let's leave it for now and let's discuss that in the future.

**Review by qooban — APPROVED** [review]

File: cypress/tsconfig.json
Comment (kubama) [reply cypress/tsconfig.json:1]: Personally I'm using auto-save, so since I'm not triggering save action explicitly I also don't format a file unless I need to, in other words unless it has style errors. I could switch to prettier, but since we are not linting JSONs there are no style errors being reported, so you shouldn't expect me reformat files that are already valid.

Comment (Gelio) [reply cypress/tsconfig.json:1]: > We already use prettier (through ESLint), maybe not in the same way

That's probably since JSON is not specified in the possible `extensions`, and that also probably requires some additional plugin :slightly_smiling_face: We could run `prettier` on files as a pre-commit hook. Also, prettier has a `--check` flag that makes it so it will fail when there are formatting differences.

Configuring it is out of scope for this PR, so I agree to tackle it in the future

**Pull request merged**
